### DISCUSSION TITLE
2월 4주차 디자인 패턴과 프로그래밍 패러다임 - 1.1.2 팩토리 패턴, 1.1.6 이터레이터 패턴

### DISCUSSION THREAD

**Discussion by choiapple** [SPHA-BORD/CS-Study#7] [📚 CS 스터디]
#6 
## 팩토리 패턴(Factory pattern)
- 객체를 사용하는 코드에서 객체 생성 부분을 떼어내 추상화한 패턴
- 상속 관계에 있는 두 클래스에서 상위 클래스가 중요한 뼈대를 결정하고, 하위 클래스에서 객체 생성에 관한 구체적인 내용을 결정하는 패턴
    - 상위 클래스와 하위 클래스가 분리되기 때문에 느슨한 결합을 가지며 상위 클래스에서는 인스턴스 생성 방식에 대해 전혀 알 필요가 없기 때문에 더 많은 유연성을 갖게된다.
    - 객체 생성 로직이 따로 떼어져 있기 때문에 코드를 리팩토링하더라도 한 곳만 고칠 수 있게 되는 유지 보수성이 증가
    - ex) 커피 공장이 있다면 여기에 원하는 커피 레시피를 넣으면 커피가 만들어진다.
- 장점 :
    - 기존 코드(인스턴스를 만드는 과정)를 수정하지 않고 새로운 인스턴스를 다른 방법으로 생성하도록 확장할 수 있다.
    - 확장에 열려있고 변경에 닫혀있는 객체지향 원칙을 적용했기 때문에 가능
- 단점 :
    - 클래스가 많아진다.
    - 제품 클래스가 바뀔 때마다 새로운 서브 클래스를 생성해야 한다.
    - 클라이언트가 creator 클래스를 반드시 상속해 product를 생성해야 한다.
##
##

 ## 이터레이터 패턴
- 이터레이터 패턴(iterator pattern) : 이터레이터(iterator)를 사용하여 컬렉션(collection)의 요소들에 접근하는 디자인 패턴
    - 순회할 수 있는 여러 가지 자료형의 구조와는 상관없이 이터레이터라는 하나의 인터페이스로 순회가 가능
    - 이터레이터 패턴을 사용하면 모든 항목에 일일이 접근하는 작업을 컬렉션 객체가 아닌 반복자 객체에서 맡게 된다. 이렇게 하면 집합체의 인터페이스 및 구현이 간단해질 뿐 아니라, 집합체에서는 반복잡업에서 손을 떼고 원래 자신이 할 일(객체 컬렉션 관리)에만 전념할 수 있다.  

##

**Comment by K-J-HYEON** [comment]
@choiapple  
**팩토리 패턴은** SOLID 원칙 중 클래스나 모듈은 확장에는 열려있어야 하고 수정에는 닫혀있어야 한다는 **개방 - 폐쇄 원칙**이 지켜진다고 들었는데, 
그와 동시에  하나의 클래스는 단 하나의 기능만을 가지는 **단일 책임 원칙** 또한 지켜진다고 보면 될까요?!! 📝

**Reply by choiapple** [reply]
팩토리 패턴은 클라이언트가 특정 객체의 생성을 직접 생성하지 않고 생성 역할을 하는 클래스를 만들어 그 클래스에게 위임합니다. 그리고 객체의 생성을 담당하는 클래스를 한 곳에서 관리하여 응집도를 높여서 단일 책임 원칙의 원칙대로 설계합니다!

**Reply by K-J-HYEON** [reply]
팩토리 패턴이 SOLID 원칙에서 SRP(Single Responsibility Principle)와 OCP(Open-Closed Principle) 둘 다 지키는군요. 팩토리 패턴의 특징인 한 클래스에서 다른 클래스의 위임과정과 그 두 클래스 간의 상속관계에서 상위클래스 역할 하위 클래스의 역할에 대해서 파악할 수 있어서 좋았습니다. 감사합니다!!
